yaml.go

```diff
@@ -11,20 +11,98 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
-// Marshal marshals the object into JSON then converts JSON to YAML and returns the
-// YAML.
-func Marshal(o interface{}) ([]byte, error) {
-	j, err := json.Marshal(o)
-	if err != nil {
-		return nil, fmt.Errorf("error marshaling into JSON: %v", err)
+// An Encoder serializes values into YAML and writes them using its io.Writer.
+type Encoder struct {
+	w io.Writer
+}
+
+// NewEncoder creates a new Encoder.
+func NewEncoder(w io.Writer) *Encoder {
+	return &Encoder{w: w}
+}
+
+// Encode serializes a value into YAML and writes it using the Encoder's
+// io.Writer.
+func (e *Encoder) Encode(o interface{}) error {
+	pipeReader, pipeWriter := io.Pipe()
+	defer pipeReader.Close()
+
+	var intermediate interface{}
+	jsonEnc := json.NewEncoder(pipeWriter)
+	// We are using yaml.Decoder here (instead of json.Decoder) because the Go
+	// JSON library doesn't try to pick the right number type (int, float,
```

Comment:
json.NewDecoder(...).UseNumber() will preserve numeric input literally (https://pkg.go.dev/encoding/json#Decoder.UseNumber) ... would that enable translating with fidelity? how does the json decoder compare to the yaml decoder in efficiency?

Reply:
also, a unit test demonstrating some of the edge cases like this would be helpful in ensuring this isn't producing user-visible changes

other examples of weird edges in yaml encoding/decoding we don't want to accidentally modify behavior on (even if the current behavior is not ultimately desired/correct):
* unquoted yaml keys of `y`: https://github.com/kubernetes/kubernetes/issues/102308
* yaml parsing of unquoted values containing all digits starting with 0 which are not octal values: https://github.com/kubernetes/kubernetes/issues/82296